### PR TITLE
Nawqaan 607 - Adding Pesticide Time Series Retrieval Web Service

### PR DESCRIPTION
**Pull request by cschroed-usgs**
Large parts of the new DAO POJO and mybatis mapper were generated using an R script. (Thanks @jiwalker-usgs!) The rest of it is pretty simple plumbing to push request params and response objects through the layers of the app. I didn't implement any tests because I didn't feel like they delivered much value on the simple plumbing aspects. Are there some worthwhile opportunities to test that I'm missing?

### CODE REVIEW

File: nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
+package gov.usgs.cida.nar.mybatis.model;
+
```
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java]: This class is an entirely generated POJO. Review can be skipped.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java [hunk 2]
```diff
@@ -6,21 +6,24 @@
...
+import gov.usgs.cida.nar.util.DateUtil;
+import java.sql.Date;
 import java.util.ArrayList;
 import java.util.List;
 import javax.ws.rs.NotFoundException;
 import org.joda.time.LocalDateTime;
 
 
 
-class PesticideSampleService implements NARService<PestSites> {
```
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: lol. I managed to botch that initial class definition pretty bad.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java [hunk 3]
```diff
@@ -34,11 +37,18 @@ public PesticideSampleService(PesticideSampleDao dao, PestSitesService sitesServ
...
 	
+		
 	@Override
-	public List<? extends PestSites> request() {
-		throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+	public List<PesticideSample> request() {
+		return request(siteQwId, constit, startDate, endDate);
 	}
 
+	public List<PesticideSample> request(List<String> siteQwId, List<String> constit, String startDateStr, String endDateStr) {
```
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: Here's an example of the plumbing. How useful is it to write a test that mocks the dao and then makes sure that the params coming into the dao's method call are of the right type? It seems like a compiler-level check, not a unit test.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/webservice/TimeSeriesWebservice.java [hunk 4]
```diff
@@ -108,6 +108,19 @@ public Response getMloadsBySite(@PathParam("siteQwId")String siteQwId,
 	
 	@GET
 	@Produces(MediaType.APPLICATION_JSON)
+	@Path("/pestsamp/site/{siteQwId}")
+	public Response getPestSampBySite(@PathParam("siteQwId")String siteQwId,
```
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/webservice/TimeSeriesWebservice.java]: more plumbing-type stuff.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/dao/PesticideSampleDao.java [hunk 5]
```diff
@@ -35,4 +38,20 @@
...
+		List<PesticideSample> result = null;
+		
+		Map<String, Object> params = new HashMap<>(11);
+		params.put(SITE_QW, siteQwId);
+		params.put(CONSTIT, constit);
+		params.put(START_DATE, startDate);
+		params.put(END_DATE, endDate);
+		
+		try (SqlSession session = sqlSessionFactory.openSession()) {
+			result = session.selectList(QUERY_PACKAGE + ".PesticideSampleMapper.getPesticideSample", params);
```
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/dao/PesticideSampleDao.java]: Is it worth writing a test that mocks the session object, and checks to see that the map being passed into sesssion.selectList() has the correct keys and values?

File: nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java (hunk 1 above)
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java]: The accessors and mutators could be succinctly tested with Derek Ashmore's [bean tester](http://www.derekashmore.com/2015/04/making-unit-testing-less-monotonous.html)(already a dependency of this project, but at that point we are simply testing the generators of the accessors and mutators, not the accessors and mutators themselves.

Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java]: Also, I think we would need to implement/generate equals and hashcode methods before the bean tester would work.

File: nar-webservices/pom.xml [hunk 6]
```diff
@@ -21,6 +21,7 @@
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<gs.version>2.3.2</gs.version>
 		<gt.version>10.1</gt.version>
+		<cobertura.version>2.7</cobertura.version>
```
Comment (duselmann) [inline nar-webservices/pom.xml]: I am glade to see coverage tools.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java (hunk 1 above)
Comment (duselmann) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java]: you primarily require hashcode when using Map which uses the hashcode. it is also a good idea (if not necessary) to implement hashcode when overriding equals. The default hashcode is based off the object pointer and like instances will not have the same hashcode unless you impl one that considers all values used in the equals function. Finally, IDEs will impl a decent version of each for us. I use Eclipse and it create fail-proof code.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java (hunk 2 above)
Comment (duselmann) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: was it auto generated or just a goof?

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java (hunk 3 above)
Comment (duselmann) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: If the compiler checks it then we do not require a test. We have no tests that check for the classnotfounderror for hard coded types. however we would want a test for IoC situations.

**Comment by duselmann** [comment]
Looks good to me. Feel free to merge. I have been OOO for a week so I will hold off in case you have need for merge control.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java (hunk 2 above)
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: Just a goof.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java (hunk 1 above)
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/mybatis/model/PesticideSample.java]: Cool. I will see if my IDE does that.

File: nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java (hunk 3 above)
Comment (cschroed-usgs) [inline nar-webservices/src/main/java/gov/usgs/cida/nar/service/PesticideSampleService.java]: Roger that!

**Comment by cschroed-usgs** [comment]
Thanks Dave. I added the equals and hashCode methods, and that was sufficient for BeanTester to like it.

**Pull request merged**